nftstorage/nft.storage
# feat: create cron nft-ttr to measure nft time to retrievability

Motivation:
* https://github.com/nftstorage/nft.storage/issues/1943

Plan

1. note all feedback
    * dont use jest for cron tests. use ava
    * dont use yargs for argv parsing, use `"sade": "^1.7.4"`
    * use less fields on nft-ttr metric [via](https://filecoinproject.slack.com/archives/C02BZPRS9HP/p1654044830889049?thread_ts=1653585353.059339&cid=C02BZPRS9HP)
    * make nft-ttr retrieval metric a histogram instead of a gauge [via](https://filecoinproject.slack.com/archives/C02BZPRS9HP/p1654055779880559?thread_ts=1653585353.059339&cid=C02BZPRS9HP)
    * alan [thoughts](https://github.com/nftstorage/nft.storage/issues/1943#issuecomment-1143650993)
        * It would be nice to be uploading some useful data rather than just junk
        * A typical NFT use case 10k drops
            * Folks typically upload a directory with 10k JSON files and also 10k of images (the files must not be the same or deduplication will occur)
        * We don't have to upload 10MB every time. In the Web3.Storage client we expose the DAG splitting size as a param to put. In NFT.Storage you could use carbites to split the DAG by a smaller size and then call storeCar multiple times.
        * When available (https://github.com/ipfs/go-ipfs/issues/8640), you might want to use the ?format=car parameter on the gateways to verify ALL the blocks of the DAG are accessible.
4. updates based on review/blocking feedback
1. merge this - whatever is required to get approval
2. prioritized other tasks to do while awaiting review approval
    * [x] update this PR to use sade for arg parsing
    * [x] update this PR such that cron package uses ava for test running (no jest)
    * [x] test retrieving from more than one gateway
6. verify once merged
    * crons are running by gh actions once the workflow is described in main branch
    * crons result in metrics successfully pushed to pushgateway
7. followup pr with
    * fixes if the aforementioned verifications fail

Possible Futures
* @alanshaw [requirement 'We should also record the time it took to upload the data.'](https://github.com/nftstorage/nft.storage/pull/1945#discussion_r892340411)
  * https://github.com/nftstorage/nft.storage/issues/1994
* upgrade tsc https://github.com/nftstorage/nft.storage/pull/1945#issuecomment-1155533297
* eslint rest of files and simplify eslint config exceptions https://github.com/nftstorage/nft.storage/pull/1945#discussion_r892271164
* [ ] change nft-ttr so it doesn't generate 10Mb every time, it just generates ~1Mb but triggers at least one car split by manually configuring the client
  * [ ] blocked by https://github.com/nftstorage/nft.storage/issues/1962 
* [ ] propose [ADRs](https://adr.github.io/)
* [ ] propose ADR in nft.storage for using sade for arg parsing
* [ ] propose ADR in nft.storage for using ava as test runner
* add test coverage requirements to cron usage of c8 (e.g. thresholds from [ucanto](https://github.com/web3-storage/ucanto/blob/main/package.json#L26))
* make this job have the nft.storage-specific parts be injectable so same job can be used to observe web3.storage
* disable eslint overrides (which were added to avoid unnecessarily modifying files otherwise not touched by this PR) in packages/cron/package.json `.eslintConfig.overrides`
* add another cronjob to continuously test @alanshaw's [thought](https://github.com/nftstorage/nft.storage/issues/1943#issuecomment-1143650993)
  > * A typical NFT use case 10k drops
  >   * Folks typically upload a directory with 10k JSON files and also 10k of images (the files must not be the same or deduplication will occur)
